File: crates/core/src/complete.rs
use crate::content::Content;
use crate::session::Session;

const NAMES: &[&str] = &[
    "ls", "cd", "cat", "pwd", "whoami", "stack", "cv", "contact", "book", "history", "help",
    "clear", "exit",
];

pub fn complete(session: &Session, content: &Content, partial: &str) -> Vec<String> {
    match partial.split_once(' ') {
        None => NAMES
            .iter()
            .filter(|name| name.starts_with(partial))
            .map(|name| name.to_string())
            .collect(),
        Some((command, rest)) => {
            let prefix = rest.trim_start();
            let Some(dir) = content.resolve_dir(&session.cwd) else {
                return Vec::new();
            };
            let mut found: Vec<String> = dir
                .dirs
                .iter()
                .map(|d| d.name.clone())
                .chain(
                    dir.files
                        .iter()
                        .filter(|f| !f.hidden)
                        .map(|f| f.name.clone()),
                )
                .filter(|name| name.starts_with(prefix) && !prefix.is_empty())
                .map(|name| format!("{command} {name}"))
                .collect();
            found.sort();
            found
        }
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::content::Content;
    use crate::session::Session;

    #[test]
    fn completes_a_command_name() {
        let content = Content::load();
        let session = Session::new();
        assert_eq!(
            complete(&session, &content, "wh"),
            vec!["whoami".to_string()]
        );
    }

    #[test]
    fn completes_a_path_argument() {
        let content = Content::load();
        let session = Session::new();
        let found = complete(&session, &content, "cd pro");
        assert!(found.contains(&"cd projects".to_string()), "got: {found:?}");
    }

    #[test]
    fn never_completes_the_hidden_file() {
        let content = Content::load();
        let session = Session::new();
        let found = complete(&session, &content, "cat .kn");
        assert!(
            found.is_empty(),
            "the egg must stay found by hand, got: {found:?}"
        );
    }

    #[test]
    fn no_match_returns_nothing() {
        let content = Content::load();
        let session = Session::new();
        assert!(complete(&session, &content, "xyz").is_empty());
    }
}

File: packages/core/src/complete.ts
import type { Content } from './content.ts'
import { resolveDir, shortcuts } from './content.ts'
import type { Session } from './session.ts'

// Aliases are left out on purpose: `quit`, `logout` and `hire` are not offered.
// Same two blocks as the help listing, in the same order.
const NAVIGATION = ['ls', 'cd', 'cat', 'pwd'] as const
const SESSION = ['cv', 'contact', 'book', 'history', 'help', 'clear', 'exit'] as const

export function complete(s: Session, c: Content, partial: string): string[] {
  const space = partial.indexOf(' ')

  // The journey's own sections complete like any other command, since that is
  // what they are. Hidden files are excluded by `shortcuts`.
  if (space === -1) {
    const names = [...NAVIGATION, ...shortcuts(c).map((f) => f.name), ...SESSION]
    return names.filter((n) => n.startsWith(partial))
  }

  const command = partial.slice(0, space)
  const prefix = partial.slice(space + 1).replace(/^\s+/, '')
  if (prefix === '') return []

  const dir = resolveDir(c, s.cwd)
  if (!dir) return []

  return [
    ...dir.dirs.map((d) => d.name),
    // The hidden file is excluded: the puzzle is found by manual exploration.
    ...dir.files.filter((f) => !f.hidden).map((f) => f.name),
  ]
    .filter((name) => name.startsWith(prefix))
    .map((name) => `${command} ${name}`)
    .sort()
}
